src/durmod/kaldi-durmod.cc

```
@@ -0,0 +1,425 @@
+    feat_idx += num_phone_identities_;
+
+    // binary features (from extra questions for eg.)
+    for (int j = 0; j < binary_feats.size(); j++) {
+      feat_elements.push_back(std::make_pair(feat_idx + binary_feats[j], 1.0));
+    }
+    feat_idx += num_binary_features_;
+
+    // duration features
+    if (i < phone_index) {
+      feat_elements.push_back(
+                             std::make_pair(feat_idx++,
+                                            NormalizeDuration(phone_duration)));
+    }
+  }
+  SparseVector<BaseFloat> tmp(feature_dim_, feat_elements);
+  feat->CopyFromSvec(tmp);
+}
+
+void PhoneDurationFeatureMaker::InitFeatureMaker(
+                                              const PhoneDurationModel &model) {
+  num_binary_features_ = model.questions_.size();
+  num_phone_identities_ = model.roots_.size() + 1;  // id=0 is for not-available
+                                                    // phones (i.e. null phones
+                                                    // which occur at edges).
+  left_context_ = model.left_context_;
+  right_context_ = model.right_context_;
+  max_duration_ = model.max_duration_;
+  int input_dim_phones = num_phone_identities_ *
+                                           (left_context_ + right_context_ + 1);
+  int input_dim_durations = left_context_;
+  int input_dim_binary = num_binary_features_ *
+                                           (left_context_ + right_context_ + 1);
+  feature_dim_ = input_dim_phones + input_dim_binary + input_dim_durations;
+
+  // create the reverse map for questions membership
+  for (int i = 0; i < model.questions_.size(); i++) {
+    for (int j = 0; j < model.questions_[i].size(); j++) {
+      int phone = model.questions_[i][j];
+      binary_feats_[phone].push_back(i);
+    }
+  }
+
+  // and the reverse map for phoneme tree roots
+  int max_phone_idx = 0;  // TODO(hhadian): in the end, check if all phones have
+                          // been categorized.
```

To actually implement this TODO I need to get the number of phones from transition model which is not available at every place that I instantiate the feature maker. I'll have to make a lot of changes. I think it would be better if we do not check at this stage, and later when making examples we can raise an error if a phone-id is not found in this map (i.e. in roots.int). Is it alright?